chain/chain/src/flat_storage_resharder.rs

```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
+                let height = chain_store.get_block_height(&flat_head_block_hash)?;
+                if height > chain_final_head.height {
+                    let msg = format!("flat head moved too far: head = {flat_head_block_hash}, height = {height}, final block height = {}", chain_final_head.height);
+                    return Err(Error::ReshardingError(msg));
+                }
+                // Stop if we reached chain final head.
+                if flat_head_block_hash == chain_final_head.last_block_hash {
+                    break;
+                }
+                flat_head_block_hash = chain_store.get_next_block_hash(&flat_head_block_hash)?;
+                if let Some(changes) = store
+                    .get_delta(shard_uid, flat_head_block_hash)
+                    .map_err(|err| Into::<StorageError>::into(err))?
+                {
+                    merged_changes.merge(changes);
+                    store_update.remove_delta(shard_uid, flat_head_block_hash);
+                }
+                // TODO (resharding): if flat_head_block_hash == state sync hash -> do snapshot
+            }
+
+            // Commit all changes to store.
+            merged_changes.apply_to_flat_state(&mut store_update, shard_uid);
+            store_update.set_flat_storage_status(
+                shard_uid,
+                FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CatchingUp(
+                    flat_head_block_hash,
+                )),
+            );
+            store_update.commit()?;
+            num_batches_done += 1;
+
+            // If we reached chain final head, we can terminate the delta application step.
+            if flat_head_block_hash == chain_final_head.last_block_hash {
+                return Ok((num_batches_done, chain_final_head));
+            }
+        }
+    }
+
+    /// Creates a flat storage entry for a shard that completed catchup. Also clears leftover data.
+    #[tracing::instrument(
+        level = "info",
+        target = "resharding",
+        "FlatStorageResharder::shard_catchup_finalize_storage",
+        skip_all,
+        fields(?shard_uid)
+    )]
+    fn shard_catchup_finalize_storage(
+        &self,
+        shard_uid: ShardUId,
+        flat_head: &Tip,
+    ) -> Result<(), Error> {
+        // GC deltas from forks which could have appeared on chain during catchup.
+        let store = self.runtime.store().flat_store();
+        let mut store_update = store.store_update();
+        // Deltas must exist because we applied them previously.
+        let deltas_metadata = store.get_all_deltas_metadata(shard_uid).unwrap_or_else(|_| {
+            panic!("Cannot read flat state deltas metadata for shard {shard_uid} from storage")
+        });
+        let mut deltas_gc_count = 0;
+        for delta_metadata in deltas_metadata {
+            if delta_metadata.block.height <= flat_head.height {
+                store_update.remove_delta(shard_uid, delta_metadata.block.hash);
```

Comment:
I would like @Longarithm to take a look at this function and check whether it looks good